### PR TITLE
`@inbounds` annotations for filter

### PR DESCRIPTION
**Pull request by andyferris**

### CODE REVIEW

**Review by nalimilan — APPROVED** [review]

**Comment by andyferris** [comment]
Is travis MacOS build busted for everyone?

File: base/array.jl [hunk 1]
```diff
@@ -2345,7 +2345,7 @@ function filter!(f, a::AbstractVector)
 
     for acurr in a
         if f(acurr)
-            a[i] = acurr
+            @inbounds a[i] = acurr
```
Comment (KristofferC) [inline base/array.jl]: Is there a policy for `@inbounds`, `AbstractArray` and the iteration protocol. I.e is it ok to segfault if someone implemented the iteration protocol erroneously?

File: base/array.jl [hunk 2]
```diff
@@ -2318,7 +2318,7 @@ julia> filter(isodd, a)
  9
 ```
 """
-filter(f, As::AbstractArray) = As[map(f, As)::AbstractArray{Bool}]
+filter(f, As::AbstractArray) = @inbounds As[map(f, As)::AbstractArray{Bool}]
```
Comment (vtjnash) [inline base/array.jl]: This seems quite unlikely to make a difference, unless our inlining heuristic is messed up and we need to go fix that instead

File: base/array.jl (hunk 1 above)
Comment (vtjnash) [reply base/array.jl]: This is iterating over the index space of a: `foreach(i -> a[i] = a[i], eachindex(a))`. I think it's OK to assume that is required to hold.

Comment (nalimilan) [reply base/array.jl]: AFAIK the policy is that `AbstractArray` implementations which opt-in to using `@boundscheck` assert that they are correct, so it's OK to crash if that's not the case. See https://github.com/JuliaLang/julia/issues/20469 and linked issues.

File: base/array.jl (hunk 2 above)
Comment (andyferris) [reply base/array.jl]: The `map` probably won’t be inlined, but I’m guessing the `getindex` may?

Obviously not a huge performance penalty but can help with tiny arrays I guess, so why not?

Comment (vtjnash) [reply base/array.jl]: Why might getindex get inlined here?

“Why not” isn’t really much of a reason to disable error checking for an unproven assertion for little expected gain

Comment (andyferris) [reply base/array.jl]: OK I've removed this.

I still have *no* idea how `@inbounds As[map(f, As)::AbstractArray{Bool}]` could possibly be unsafe, unless someone completely borks up an implementation of `map`, but let's merge the other more useful change in this PR :)

**Comment by andyferris** [comment]
Should performance improvements like this one be backported?

**Comment by KristofferC** [comment]
Yes

**Pull request merged**